### PR TITLE
feature: classless

### PR DESCRIPTION
**Pull request by raveclassic**
This PR:
 - replaces classes with top-level functions
 - simplifies RemoteData codec (io-ts)
 - contains many **breaking changes**

### CODE REVIEW

File: src/remote-data.ts [hunk 1]
```diff
@@ -29,879 +30,155 @@ declare module 'fp-ts/lib/HKT' {
...
+};
 
 /**
  * Represents a value of one of four possible types (a disjoint union)
  *
- * An instance of `RemoteData` is either an instance of `RemoteInitial`, `RemotePending`, `RemoteFailure` or `RemoteSuccess`
+ * An instance of {@link RemoteData} is either an instance of {@link RemoteInitial}, {@link RemotePending}, {@link RemoteFailure} or {@link RemoteSuccess}
  *
- * A common use of `RemoteData` is as an alternative to `Either` or `Option` supporting initial and pending states (fits best with [RXJS]{@link https://github.com/ReactiveX/rxjs/}).
+ * A common use of {@link RemoteData} is as an alternative to `Either` or `Option` supporting initial and pending states (fits best with [RXJS]{@link https://github.com/ReactiveX/rxjs/}).
```
[the remark this answers is not in the document]
Comment (raveclassic) [reply src/remote-data.ts]: Yeah, I didn't dig into docs that much, just tweaked type references to be picked by WS more accurately. I'm for moving overall documentation to another ticket

File: src/remote-data.ts [hunk 2]
```diff
@@ -29,879 +30,155 @@ declare module 'fp-ts/lib/HKT' {
...
+/**
+ * Takes a default value as an argument.
+ * If this {@link RemoteData} is "Left" part it will return default value.
+ * If this {@link RemoteData} is {@link RemoteSuccess} it will return it's value ("wrapped" value, not default value)
+ *
+ * Note: Default value should be the same type as {@link RemoteData} (internal) value, if you want to pass different type as default, use {@link fold}.
+ *
+ * @example
+ * getOrElse(() => 999)(some(1)) // 1
+ * getOrElseValue(() => 999)(initial) // 999
```
[the remark this answers is not in the document]
Comment (raveclassic) [reply src/remote-data.ts]: There's no strict `getOrElse` and lazy `getOrElseL` in `fp-ts@2.0` for the sake of API surface minimization, only lazy `getOrElse`. Anyway doc improvements can be moved to separate ticket

File: src/__tests__/remote-data.spec.ts [hunk 3]
```diff
@@ -34,411 +48,478 @@ describe('RemoteData', () => {
...
+					const double = (a: number): number => a * 2;
+					const quad = flow(
+						double,
+						double,
+					);
+					it('initial', () => {
+						expect(remoteData.map(initial, quad)).toBe(
+							pipe(
+								initial,
+								RD.map(double),
```
[the remark this answers is not in the document]
Comment (raveclassic) [reply src/__tests__/remote-data.spec.ts]: Yeah, me too. I changed only the parts related to `fp-ts@2.0`. Should we do refactoring here?

[the remark this answers is not in the document]
Comment (sutarmin) [reply src/__tests__/remote-data.spec.ts]: As it should be discussed before doing anything, I'm for separating this into another ticket

**Comment by sutarmin** [comment]
> should we consider supporting `RemoteData` for `fp-ts@1`? Because for _some_ projects it's quite difficult (not to say impossible) to migrate to `fp-ts@2`.

I hope you did not ignore this :)

**Comment by raveclassic** [comment]
> > should we consider supporting `RemoteData` for `fp-ts@1`? Because for _some_ projects it's quite difficult (not to say impossible) to migrate to `fp-ts@2`.
> 
> I hope you did not ignore this :)

Preferred way is to follow `fp-ts` support schedule. I'm ok to support `remote-data-ts@1.x` until `fp-ts@1.x` is supported.

**Pull request merged**